problem/0000-metrics-v1.md

```diff
@@ -0,0 +1,136 @@
+# Linkerd Metrics v1
+
+- Contribution Name: metrics-v1
+- Implementation Owner:
+- Start Date: 2020-04-23
+- Target Date:
+- RFC PR:
+- Linkerd Issue:
+- Reviewers: @grampelberg @adleong
+
+## Problem Statement
+
+[problem-statement]: #problem-statement
+
+Linkerd's `stat` API powers both the command-line and web interfaces, but it's tailored closely
+to their use cases and it has evolved only to express the features of these tools. Because of its
+narrow focus, integrators tend to read data directly from Prometheus, which features an
+expressive query language and a variety of external integrations/extensions. This presents an
+implicit, undocumented relationship between the proxy's metrics exposition structure and traffic
+in Kubernetes.
+
+The goal of this RFC is to define an external-facing API for Linkerd that:
+
+1. The semantics of which are defined by this RFC;
+2. Is stable/versioned;
+3. Supports Linkerd's UIs as well as arbitrary external clients;
+4. Supports Kubernetes RBAC;
+5. Supports extensions for additional protocols, metrics, and metadata; and
+6. Supports arbitrary workload hierarchies (i.e for custom operators).
+
+## Design Notes
+
+### API
+
+The metrics API is read-only.
+
+In order to support RBAC, all requests should be scoped to the resources on which metrics are
+measured.
+
+It is time-oriented. It should have both time-series and summary capabilities.
```

Comment:
I think so. Even though linkerd's interfaces only expose summary data, it seems extremely desirable to expose timeseries. Otherwise, we're basically forcing all of these needs through the raw underlying prometheus interface. Though, from the impl-plan point of view I think we could avoid doing this right away, I think we should plan for the API to satisfy these needs.

Reply:
For what it's worth, when you have a chain of federated Prometheuses, each Prometheus exposes point-in-time data to the next Prometheus in the chain which scrapes it at regular intervals.  So it's not unheard of.

I'm not sure if I'm reasoning about this correctly, but I think if you have mismatched scrape intervals, the resolution of the data will effectively be the longest scrape interval.  I don't think the intervals will compound in some wonky way.

That said, it seems to me like there are 3 different formats our API could expose metrics in: 
* summaries over a time period (as the gRPC API does today)
* point-in-time snapshot of gauges/counters/histograms (ala hitting the /metrics or /federate endpoint)
* timeseries (like what is returned from a promQL query)

We need to decide which of these we want to support.

Reply:
@adleong That all makes sense to me.

I think that if we can support the generic time-series case, the summary case becomes ~trivial (it's a `sum` or `rate` query over the interval, right?).

Reply:
Yeah, assuming the API supports this (as promQL does).